internal/mesh/internal/controllers/sidecarproxy/controller.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package sidecarproxy
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/cache/sidecarproxycache"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecarproxy/builder"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecarproxy/fetcher"
+	"github.com/hashicorp/consul/internal/mesh/internal/mappers/sidecarproxymapper"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+// ControllerName is the name for this controller. It's used for logging or status keys.
+const ControllerName = "consul.io/sidecar-proxy-controller"
+
+type TrustDomainFetcher func() (string, error)
+
+func Controller(cache *sidecarproxycache.Cache, mapper *sidecarproxymapper.Mapper, trustDomainFetcher TrustDomainFetcher) controller.Controller {
+	if cache == nil || mapper == nil || trustDomainFetcher == nil {
+		panic("cache, mapper and trust domain fetcher are required")
+	}
+
+	return controller.ForType(types.ProxyStateTemplateType).
+		WithWatch(catalog.ServiceEndpointsType, mapper.MapServiceEndpointsToProxyStateTemplate).
+		WithWatch(types.UpstreamsType, mapper.MapDestinationsToProxyStateTemplate).
+		WithReconciler(&reconciler{cache: cache, getTrustDomain: trustDomainFetcher})
+}
+
+type reconciler struct {
+	cache          *sidecarproxycache.Cache
+	getTrustDomain TrustDomainFetcher
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	rt.Logger = rt.Logger.With("resource-id", req.ID, "controller", ControllerName)
+
+	rt.Logger.Trace("reconciling proxy state template", "id", req.ID)
+
+	// Instantiate a data fetcher to fetch all reconciliation data.
+	dataFetcher := fetcher.Fetcher{Client: rt.Client, Cache: r.cache}
+
+	// Check if the workload exists.
+	workloadID := resource.ReplaceType(catalog.WorkloadType, req.ID)
+	workload, err := dataFetcher.FetchWorkload(ctx, resource.ReplaceType(catalog.WorkloadType, req.ID))
+	if err != nil {
+		rt.Logger.Error("error reading the associated workload", "error", err)
+		return err
+	}
+	if workload == nil {
```

Comment:
Should you delete any ProxyStateTemplates because of the workload going away?

Reply:
No because we should generate a new one if workload exists but proxy state template doesn't